pkg/controller/operators/olm/initializer/inject.go

```diff
@@ -0,0 +1,151 @@
+package initializer
+
+import (
+	"errors"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// InjectEnvIntoDeployment injects the proxy env variables specified in
+// proxyEnvVar into the container(s) of the given PodSpec.
+//
+// If any Container in PodSpec already defines an env variable of the same name
+// as any of the proxy env variables then it
+func InjectEnvIntoDeployment(podSpec *corev1.PodSpec, envVars []corev1.EnvVar) error {
+	if podSpec == nil {
+		return errors.New("no pod spec provided")
+	}
+
+	for i := range podSpec.Containers {
+		container := &podSpec.Containers[i]
+		container.Env = mergeEnvVars(container.Env, envVars)
+	}
+
+	return nil
+}
+
+func mergeEnvVars(containerEnvVars []corev1.EnvVar, newEnvVars []corev1.EnvVar) (merged []corev1.EnvVar) {
+	merged = containerEnvVars
+
+	for _, newEnvVar := range newEnvVars {
+		existing, found := findEnvVar(containerEnvVars, newEnvVar.Name)
+		if found {
+			existing.Value = newEnvVar.Value
+			continue
+		}
+
+		merged = append(merged, corev1.EnvVar{
+			Name:  newEnvVar.Name,
+			Value: newEnvVar.Value,
+		})
+	}
+
+	return
+}
+
+func findEnvVar(proxyEnvVar []corev1.EnvVar, name string) (envVar *corev1.EnvVar, found bool) {
+	for i := range proxyEnvVar {
+		if name == proxyEnvVar[i].Name {
+			// Environment variable names are case sensitive.
+			found = true
+			envVar = &proxyEnvVar[i]
+
+			break
+		}
+	}
+
+	return
+}
+
+// InjectVolumesIntoDeployment injects the provides Volumes
+// into the container(s) of the given PodSpec.
+//
+// If any Container in PodSpec already defines a Volume of the same name
+// as any of the provided Volumes then it will be overwritten.
+func InjectVolumesIntoDeployment(podSpec *corev1.PodSpec, volumes []corev1.Volume) error {
+	if podSpec == nil {
+		return errors.New("no pod spec provided")
+	}
+
+ 	podSpec.Volumes = mergeVolumes(podSpec.Volumes, volumes)
+
+	return nil
+}
+
+func mergeVolumes(podSpecVolumes []corev1.Volume, newVolumes []corev1.Volume) (merged []corev1.Volume) {
+	merged = podSpecVolumes
+
+	for _, newVolume := range newVolumes {
+		existing, found := findVolume(podSpecVolumes, newVolume.Name)
+		if found {
+			*existing = newVolume
+			continue
+		}
+
+		merged = append(merged, newVolume)
+	}
+
+	return
+}
+
+func findVolume(volumes []corev1.Volume, name string) (envVar *corev1.Volume, found bool) {
+	for i := range volumes {
+		if name == volumes[i].Name {
+			// Environment variable names are case sensitive.
+			found = true
+			envVar = &volumes[i]
+
+			break
+		}
+	}
+
+	return
+}
+
+// InjectVolumeMountsIntoDeployment injects the provides VolumeMounts
+// into the given PodSpec.
+//
+// If the PodSpec already defines a VolumeMount of the same name
+// as any of the provided VolumeMounts then it will be overwritten.
+func InjectVolumeMountsIntoDeployment(podSpec *corev1.PodSpec, volumeMounts []corev1.VolumeMount) error {
```

Comment:
Should we check that that `Volume`s referenced by the `volumeMounts` exist?

Reply:
Sure thing!

Reply:
@ecordell and I discussed this offline and have decided not to include this feature in this PR. Instead, we are going to investigated validating the subscription upfront with an admission webhook for Subscriptions